### PR TITLE
Reproduce S2583 FP: indexer does not throw exception

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**
Reproduces #9580

### CODE REVIEW

**Review by sebastien-marichal — APPROVED** [review]
LGTM!
I will let @Tim-Pohlmann to merge this as I might miss some SE knowledge.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs [hunk 1]
```diff
@@ -4310,3 +4311,20 @@ void Test(List<Repro9425> aList)
...
+        if (element != null)
+        {
+            Console.WriteLine(element.ToString());
+        }
+        if (collection.Count == 0)          // Noncompliant - FP: the indexer with string argument returns null if the key is not found rather than throwing an exception,
+        {                                   // so at this point we can't know for sure that the collection is not empty.
+            Console.WriteLine("Empty!");    // Secondary - FP
+        }
+    }
+}
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:4330]: This is an FP indeed (I didn't knew this behavior exists, this is super inconsistent...).
However, I do think the problem in #9580 is the call to [Refresh](https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.replication.replicationbasecollection.refresh?view=sqlserver-2016#microsoft-sqlserver-replication-replicationbasecollection-refresh).

The underlying issue is, that we need to be more careful with our learnings if the collection in question is not of a type the Symbolic Execution engine is aware of.

I'd make two reproducers:
1. The one you have here with `NameValueCollection`. Add a second test case with a custom collection that overrides the indexer.
2. A reproducer with a custom collection that has a custom method to add items to the collection (like `Refresh`). Please check before-hand if a UT for a case like this exists already. Especially take a look at the S4158 test cases, there might be something in there.

**Review by Tim-Pohlmann — COMMENTED** [review]
This FP is a good find but I do not think it reproduces the original issue.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs (hunk 1 above)
Comment (zsolt-kolbay-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/ConditionEvaluatesToConstant.cs:4330]: https://github.com/SonarSource/sonar-dotnet/pull/9592

**Pull request merged**